qiskit/dagcircuit/dagcircuit.py

```diff
@@ -977,6 +977,57 @@ def topological_op_nodes(self, key=None):
         """
         return (nd for nd in self.topological_nodes(key) if isinstance(nd, DAGOpNode))
 
+    def replace_block_with_op(self, node_block, op, wire_pos_map):
```

Comment:
Yeah looking at this locally doing the checking can hurt performance. I tested both the path length check and just adding `is_directed_acyclic_graph()` after the operations. Adding a cycle check at the end makes this ~10x slower in my testing while doing the distance approach is ~100x slower (there might be room for improvement here because I just called dijkstra's in for loop). The tradeoff is that the distance approach is much cleaner code as it's an upfront check on the inputs.

After all this I agree with @georgios-ts and @kdk and think that documenting this is probably enough. I'm thinking for this case though making the check an opt-out documenting the tradeoff between performance and validation and then for consolidate blocks we explicitly opt-out to avoid the overhead.

Reply:
A plain dijkstra in the original DAG would not actually work here because if we contract a sequence of nodes `u_i` we should not follow edges with both endpoints in the sequence while searching for a path. As an improved solution, I can't help but advertise the pending BFS/DFS PR's in retworkx that might come in handy here and avoid the need to modify the original DAG. In any case, let's stick with `is_directed_acyclic_graph()` check for now and we can revisit this.